File: packages-gen/js/lang/common-lib.js
function type_native(obj){
  if(obj == null){
    return null;
  }
  let t = typeof obj;
  if(t == "object"){
    if(Array.isArray(obj)){
      return "array";
    }
    else{
      let tn = obj["constructor"]["name"];
      if(tn == "Object"){
        return "object";
      }
      else{
        return tn;
      }
    }
  }
  else{
    return t;
  }
}

function type_class(x){
  let ntype = null;
  if(x == null){
    return null;
  }
  let t = typeof x;
  if(t == "object"){
    if(Array.isArray(x)){
      ntype = "array";
    }
    else{
      let tn = x["constructor"]["name"];
      if(tn == "Object"){
        ntype = "object";
      }
      else{
        ntype = tn;
      }
    }
  }
  else{
    ntype = t;
  }
  if((null != x) && ("object" == (typeof x)) && !Array.isArray(x)){
    return (null == x["::"]) ? "object" : x["::"];
  }
  else{
    return ntype;
  }
}

function to_string(x){
  return String(x);
}

function to_number(x){
  return Number(x);
}

function nilp(x){
  return null == x;
}

function not_nilp(x){
  return null != x;
}

function is_booleanp(x){
  return "boolean" == (typeof x);
}

function is_integerp(x){
  return Number.isInteger(x);
}

function is_numberp(x){
  return "number" == (typeof x);
}

function is_stringp(x){
  return "string" == (typeof x);
}

function is_functionp(x){
  return "function" == (typeof x);
}

function is_arrayp(x){
  return Array.isArray(x);
}

function is_objectp(x){
  return (null != x) && ("object" == (typeof x)) && !Array.isArray(x);
}

function noop(){
  return null;
}

function identity(x){
  return x;
}

function T(x){
  return true;
}

function F(x){
  return false;
}

function add(a,b){
  return a + b;
}

function sub(a,b){
  return a - b;
}

function mul(a,b){
  return a * b;
}

function div(a,b){
  return a / b;
}

function gt(a,b){
  return a > b;
}

function lt(a,b){
  return a < b;
}

function gte(a,b){
  return a >= b;
}

function lte(a,b){
  return a <= b;
}

function eq(a,b){
  return a == b;
}

function neq(a,b){
  return a != b;
}

function neg(x){
  return -x;
}

function inc(x){
  return x + 1;
}

function dec(x){
  return x - 1;
}

function zerop(x){
  return x == 0;
}

function posp(x){
  return x > 0;
}

function negp(x){
  return x < 0;
}

function evenp(x){
  return 0 == (x % 2);
}

function oddp(x){
  return !(0 == (x % 2));
}

function wrap_callback(callbacks,key){
  if(null == callbacks){
    callbacks = {};
  }
  let result_fn = function (result){
    let f = callbacks[key];
    if(null != f){
      return f.apply(null,[result]);
    }
    else{
      return result;
    }
  };
  return result_fn;
}

function return_encode(out,id,key){
  let type_fn = function (obj){
    if(obj == null){
      return null;
    }
    let t = typeof obj;
    if(t == "object"){
      if(Array.isArray(obj)){
        return "array";
      }
      else{
        let tn = obj["constructor"]["name"];
        if(tn == "Object"){
          return "object";
        }
        else{
          return tn;
        }
      }
    }
    else{
      return t;
    }
  };
  let ts = type_fn(out);
  if("function" == ts){
    return JSON.stringify({
      "id":id,
      "key":key,
      "type":"raw",
      "return":"function",
      "value":out.toString()
    });
  }
  else if("object" != ts){
    return JSON.stringify({"id":id,"key":key,"type":"data","return":ts,"value":out});
  }
  else if(null == out){
    return JSON.stringify(
      {"id":id,"key":key,"type":"data","return":"nil","value":out}
    );
  }
  else{
    try{
      return JSON.stringify({"id":id,"key":key,"type":"data","return":ts,"value":out});
    }
    catch(e){
      return JSON.stringify({
        "id":id,
        "key":key,
        "type":"raw",
        "return":ts,
        "value":out.toString()
      });
    }
  }
}

function return_wrap(f){
  try{
    let out = f();
    return return_encode(out);
  }
  catch(e){
    let err = ("string" == (typeof e)) ? e : {"message":e["message"],"stack":e["stack"]};
    return JSON.stringify({"type":"error","value":err})
  }
}

function return_eval(s){
  return return_wrap(function (){
    return eval(s);
  });
}

module.exports = {
  ["type_native"]:type_native,
  ["type_class"]:type_class,
  ["to_string"]:to_string,
  ["to_number"]:to_number,
  ["nilp"]:nilp,
  ["not_nilp"]:not_nilp,
  ["is_booleanp"]:is_booleanp,
  ["is_integerp"]:is_integerp,
  ["is_numberp"]:is_numberp,
  ["is_stringp"]:is_stringp,
  ["is_functionp"]:is_functionp,
  ["is_arrayp"]:is_arrayp,
  ["is_objectp"]:is_objectp,
  ["noop"]:noop,
  ["identity"]:identity,
  ["T"]:T,
  ["F"]:F,
  ["add"]:add,
  ["sub"]:sub,
  ["mul"]:mul,
  ["div"]:div,
  ["gt"]:gt,
  ["lt"]:lt,
  ["gte"]:gte,
  ["lte"]:lte,
  ["eq"]:eq,
  ["neq"]:neq,
  ["neg"]:neg,
  ["inc"]:inc,
  ["dec"]:dec,
  ["zerop"]:zerop,
  ["posp"]:posp,
  ["negp"]:negp,
  ["evenp"]:evenp,
  ["oddp"]:oddp,
  ["wrap_callback"]:wrap_callback,
  ["return_encode"]:return_encode,
  ["return_wrap"]:return_wrap,
  ["return_eval"]:return_eval
}

File: packages-gen/dart/lang/lib/common-lib.dart
type_native(obj) {
  if(obj == null){
    return null;
  }
  var rtype41101 = (obj.runtimeType).toString();
  var sval41102 = (obj).toString();
  if("String" == rtype41101){
    return "string";
  }
  if(("int" == rtype41101) || ("double" == rtype41101) || ("num" == rtype41101)){
    return "number";
  }
  if("bool" == rtype41101){
    return "boolean";
  }
  if(rtype41101.contains("Function") || rtype41101.contains("=>") || sval41102.startsWith("Closure")){
    return "function";
  }
  if(("List" == rtype41101) || rtype41101.contains("List")){
    return "array";
  }
  if(("Map" == rtype41101) || rtype41101.contains("Map")){
    return "object";
  }
  return rtype41101.toLowerCase();
}

type_class(x) {
  var ntype = ((value) {
    if(value == null){
      return null;
    }
    var rtype41139 = (value.runtimeType).toString();
    var sval41140 = (value).toString();
    if("String" == rtype41139){
      return "string";
    }
    if(("int" == rtype41139) || ("double" == rtype41139) || ("num" == rtype41139)){
      return "number";
    }
    if("bool" == rtype41139){
      return "boolean";
    }
    if(rtype41139.contains("Function") || rtype41139.contains("=>") || sval41140.startsWith("Closure")){
      return "function";
    }
    if(("List" == rtype41139) || rtype41139.contains("List")){
      return "array";
    }
    if(("Map" == rtype41139) || rtype41139.contains("Map")){
      return "object";
    }
    return rtype41139.toLowerCase();
  })(x);
  if(("Map" == (x.runtimeType).toString()) || (x.runtimeType).toString().startsWith("_Map") || (x.runtimeType).toString().startsWith("LinkedMap")){
    return (null == x["::"]) ? "object" : x["::"];
  }
  else{
    return ntype;
  }
}

to_string(x) {
  return (x).toString();
}

to_number(x) {
  return num.parse(x);
}

nilp(x) {
  return null == x;
}

not_nilp(x) {
  return null != x;
}

is_booleanp(x) {
  return "bool" == (x.runtimeType).toString();
}

is_integerp(x) {
  return "int" == (x.runtimeType).toString();
}

is_numberp(x) {
  return ("int" == (x.runtimeType).toString()) || ("double" == (x.runtimeType).toString()) || ("num" == (x.runtimeType).toString());
}

is_stringp(x) {
  return "String" == (x.runtimeType).toString();
}

is_functionp(x) {
  return (x.runtimeType).toString().contains("Function") || (x.runtimeType).toString().contains("=>") || (x).toString().startsWith("Closure");
}

is_arrayp(x) {
  return (x.runtimeType).toString().startsWith("List") || (x.runtimeType).toString().startsWith("_GrowableList");
}

is_objectp(x) {
  return ("Map" == (x.runtimeType).toString()) || (x.runtimeType).toString().startsWith("_Map") || (x.runtimeType).toString().startsWith("LinkedMap");
}

noop() {
  return null;
}

identity(x) {
  return x;
}

T(x) {
  return true;
}

F(x) {
  return false;
}

add(a, b) {
  return a + b;
}

sub(a, b) {
  return a - b;
}

mul(a, b) {
  return a * b;
}

div(a, b) {
  return a / b;
}

gt(a, b) {
  return (("String" == (a.runtimeType).toString()) || ("String" == (b.runtimeType).toString())) ? ((a).toString().compareTo((b).toString()) > 0) : (a > b);
}

lt(a, b) {
  return (("String" == (a.runtimeType).toString()) || ("String" == (b.runtimeType).toString())) ? ((a).toString().compareTo((b).toString()) < 0) : (a < b);
}

gte(a, b) {
  return (("String" == (a.runtimeType).toString()) || ("String" == (b.runtimeType).toString())) ? ((a).toString().compareTo((b).toString()) >= 0) : (a >= b);
}

lte(a, b) {
  return (("String" == (a.runtimeType).toString()) || ("String" == (b.runtimeType).toString())) ? ((a).toString().compareTo((b).toString()) <= 0) : (a <= b);
}

eq(a, b) {
  return a == b;
}

neq(a, b) {
  return a != b;
}

neg(x) {
  return -x;
}

inc(x) {
  return x + 1;
}

dec(x) {
  return x - 1;
}

zerop(x) {
  return x == 0;
}

posp(x) {
  return x > 0;
}

negp(x) {
  return x < 0;
}

evenp(x) {
  return 0 == (x % 2);
}

oddp(x) {
  return !(0 == (x % 2));
}

wrap_callback(callbacks, key) {
  if(null == callbacks){
    callbacks = <dynamic, dynamic>{};
  }
  var result_fn = (result) {
    var f = callbacks[key];
    if(null != f){
      return Function.apply(f,<dynamic>[result]);
    }
    else{
      return result;
    }
  };
  return result_fn;
}

return_encode(out, id, key) {
  if(out == null){
    return json.encode(
      <dynamic, dynamic>{"id":id,"key":key,"type":"data","return":"nil","value":null}
    );
  }
  var outtype41177 = (out.runtimeType).toString();
  var outstr41178 = (out).toString();
  var compact41179 = (() {
     final omitNilKeys = <String>{"db/remove", "select_method", "select_control", "return_method", "return_query", "return_count", "return_id", "return_bulk", "return_omit", "data_only", "model_id", "view_id"};
     dynamic compact(dynamic value) {
       if (value is Function) {
         return null;
       }
       if (value is Map) {
         final out = <dynamic, dynamic>{};
         value.forEach((key, entry) {
           if (!(entry is Function)) {
             final next = compact(entry);
             if (next != null || !(key is String && omitNilKeys.contains(key))) {
               out[key] = next;
             }
           }
         });
         return out;
       }
       if (value is List) {
         return List<dynamic>.from(value.where((entry) => !(entry is Function))
                                          .map((entry) => compact(entry)));
       }
       return value;
     }
     return compact( out );
   })();
  if("String" == outtype41177){
    return json.encode(
      <dynamic, dynamic>{"id":id,"key":key,"type":"data","return":"string","value":out}
    );
  }
  if(("int" == outtype41177) || ("double" == outtype41177) || ("num" == outtype41177)){
    return json.encode(
      <dynamic, dynamic>{"id":id,"key":key,"type":"data","return":"number","value":out}
    );
  }
  if("bool" == outtype41177){
    return json.encode(
      <dynamic, dynamic>{"id":id,"key":key,"type":"data","return":"boolean","value":out}
    );
  }
  if(("List" == outtype41177) || outtype41177.contains("List")){
    return json.encode(<dynamic, dynamic>{
      "id":id,
      "key":key,
      "type":"data",
      "return":"array",
      "value":compact41179
    });
  }
  if(("Map" == outtype41177) || outtype41177.contains("Map")){
    return json.encode(<dynamic, dynamic>{
      "id":id,
      "key":key,
      "type":"data",
      "return":"object",
      "value":compact41179
    });
  }
  if(outtype41177.contains("Function") || outtype41177.contains("=>") || outstr41178.startsWith("Closure")){
    return json.encode(<dynamic, dynamic>{
      "id":id,
      "key":key,
      "type":"raw",
      "return":"function",
      "value":outstr41178
    });
  }
  return json.encode(<dynamic, dynamic>{
    "id":id,
    "key":key,
    "type":"raw",
    "return":outtype41177.toLowerCase(),
    "value":outstr41178
  });
}

return_wrap(f) {
  try{
    var out = f();
    try{
      return Function.apply(return_encode,<dynamic>[out]);
    }
    catch(_){
      return Function.apply(return_encode,<dynamic>[out,null,null]);
    }
  }
  catch(e){
    return json.encode(<dynamic, dynamic>{
      "type":"error",
      "value":<dynamic, dynamic>{"message":e.toString()}
    });
  }
}

return_eval(s) {
  throw "eval not supported in Dart";
}